ibis/backends/base/connection.py

```diff
@@ -0,0 +1,187 @@
+from __future__ import annotations
+
+import abc
+from typing import Callable, List, Optional
+
+import ibis.expr.schema as sch
+import ibis.expr.types as ir
+from ibis.common.exceptions import TranslationError
+
+
+class BaseConnection(abc.ABC):
+    """
+    Base class for all backend connections.
+
+    All Ibis backends are expected to subclass this `Connection` class,
+    and implement all the required methods.
+    """
+
+    @property
+    @abc.abstractmethod
+    def name(self) -> str:
+        """
+        Name of the backend, for example 'sqlite'.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def kind(self) -> str:
+        """
+        Backend kind. One of:
+
+        sqlalchemy
+            Backends using a SQLAlchemy dialect.
+        sql
+            SQL based backends, not based on a SQLAlchemy dialect.
+        pandas
+            Backends using pandas to store data and perform computations.
+        spark
+            Spark based backends.
+        """
+        pass
+
+    @abc.abstractmethod
+    def connect(connection_string: str, **options) -> BaseConnection:
+        """
+        Connect to the underlying database and return a client object.
+        """
+        pass
+
+    def register_options(self) -> None:
+        """
+        If the backend has custom options, register them here.
+        They will be prefixed with the name of the backend.
+        """
+        pass
+
+    def compile(self, expr: ir.Expr, params=None) -> str:
```

Comment:
I think only SQL backend would return a str for this. So many should change this to `Any` in base class.